File: dp_primer/dp_primer_lis_boss.cs
// 実行時間 190ms
using System;
using static System.Linq.Enumerable;

internal class Program
{
    static void Main()
    {
        int n = int.Parse( Console.ReadLine() );
        int o = 1;
        int[] a = new int[n];
        int[] p = new int[n];

        foreach( int i in Range( 0, n ) ) {
            a[i] = int.Parse( Console.ReadLine() );
        }
        p[0] = 1;
        foreach( int i in Range( 1, n - 1 ) ) {
            p[i] = 1;
            foreach( int j in Range( 0, i ) ) {
                if( a[i] < a[j] ) {
                    p[i] = Math.Max( p[i], p[j] + 1 );
                    o = Math.Max( o, p[i] );
                }
            }
        }
        Console.WriteLine( o );
        return;
    }
}

File: dp_primer/dp_primer_lis_boss.py
# 実行時間 3480ms
import sys

rl = sys.stdin.readline
n = int(rl())
o = 1
a = []
p = [1]

for _ in [0] * n:
    a.append(int(rl()))
for i in range(1, n):
    p.append(1)
    for j in range(i):
        if a[i] < a[j]:
            if p[i] < p[j] + 1:
                p[i] = p[j] + 1
            if o < p[i]:
                o = p[i]
print(o)
